driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -663,20 +680,35 @@ private boolean addConnectionIfUnderMaximum(int shardId, ConnectionTasksSharedSt
           open[shardId].decrementAndGet();
           return false;
         }
-        logger.debug("Creating new connection on busy pool to {}", host);
-        newConnection = sharedState.registerTask(shardId);
+        newConnection = sharedState.getConnection(shardId);
         if (newConnection == null) {
-          try {
-            do {
-              newConnection = manager.connectionFactory().open(this);
-              if (newConnection.shardId() == shardId) {
-                newConnection.setKeyspace(manager.poolsState.keyspace);
-              } else {
-                newConnection = sharedState.addConnectionToClose(shardId, newConnection);
-              }
-            } while (newConnection == null);
-          } finally {
-            sharedState.unregisterTask();
+          int shardCount = 1;
+          int localPort = 0;
+          InetSocketAddress serverAddress = host.getEndPoint().resolve();
+          int serverPort = serverAddress.getPort();
+          ShardingInfo shardingInfo = host.getShardingInfo();
+          if (shardingInfo != null) {
+            shardCount = shardingInfo.getShardsCount();
+            if (shardingInfo.getShardAwarePort() != 0) {
+              serverPort = shardingInfo.getShardAwarePort();
+              localPort = getAvailablePortForShard(shardCount, shardId, 100);
+            }
```

Comment:
Imho it is a bad idea to split the port finding and the connection creation, since it is rather possible someone snags the port you deviced between the check socket creation+close and the client open below. 
Why not just abstract the connection open into an interface and use one that loop-tries shard ports for this case, and a straighforward one for non-shard?